### PR TITLE
Port Mask R-CNN to Keras3

### PR DESCRIPTION
**Pull request by laxmareddyp**
# What does this PR do?

Port legacy Mask R-CNN to Keras 3

Please see the attached demo [gist](https://colab.sandbox.google.com/gist/laxmareddyp/f7c8196338394a7f9c498658931b292f/mask-rcnn_final.ipynb#scrollTo=F-vSdTWve4EG) for reference.

### CODE REVIEW

File: keras_cv/src/layers/object_detection/non_max_suppression.py [hunk 1]
```diff
@@ -63,7 +63,12 @@ def __init__(
         self.built = True
 
     def call(
-        self, box_prediction, class_prediction, images=None, image_shape=None
```
Comment (divyashreepathihalli) [inline keras_cv/src/layers/object_detection/non_max_suppression.py:66]: Let us copy this file over into the mask_rcnn folder and make the changes there

File: keras_cv/src/layers/object_detection/roi_sampler.py [hunk 2]
```diff
@@ -44,7 +45,7 @@ class ROISampler(keras.layers.Layer):
       roi_bounding_box_format: The format of roi bounding boxes. Refer
         [to the keras.io docs](https://keras.io/api/keras_cv/bounding_box/formats/)
         for more details on supported bounding box formats.
-      gt_bounding_box_format: The format of ground truth bounding boxes.
+      gt_bounding_box_format: The format of gt bounding boxes.
```
Comment (divyashreepathihalli) [inline keras_cv/src/layers/object_detection/roi_sampler.py:48]: expand gt->ground truth

File: keras_cv/src/layers/object_detection/roi_sampler.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (divyashreepathihalli) [inline keras_cv/src/layers/object_detection/roi_sampler.py:13]: copy the roi sampler file to the model folder and then make changes to it

File: keras_cv/src/layers/object_detection/roi_sampler.py [hunk 4]
```diff
@@ -56,6 +57,7 @@ class ROISampler(keras.layers.Layer):
         further (loss) calculation, defaults to 256.
       append_gt_boxes: boolean, whether gt_boxes will be appended to rois
         before sample the rois, defaults to True.
+      mask_shape: The shape of segmentation masks used for training.
```
Comment (divyashreepathihalli) [inline keras_cv/src/layers/object_detection/roi_sampler.py:60]: mention default value

File: keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py [hunk 5]
```diff
@@ -292,6 +292,7 @@ def __init__(
         rois = _clip_boxes(rois, "yxyx", image_shape)
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/faster_rcnn/faster_rcnn.py:293]: we could add `faster_rcnn_backbone` in mask_rcnn folder and make these changes there

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+@keras_cv_export(
+    "keras_cv.models.faster_rcnn.MaskHead",
+    package="keras_cv.models.faster_rcnn",
+)
+class MaskHead(keras.layers.Layer):
+    """A Keras layer implementing the R-CNN Mask Head.
+
+    The architecture is adopted from Matterport's Mask R-CNN implementation.
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:26]: add link to the ref

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    Args:
+        num_classes: The number of object classes that are being detected.
+        conv_dims: (Optional) a list of integers specifying the number of
+            filters for each convolutional layer. Defaults to [].
+        deconv_dim: (Optional) the numver of filters to use in the upsampling
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:32]: nit: numver->number

Comment (divyashreepathihalli) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:32]: mention default values

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                deconv_dim,
+                kernel_size=2,
+                strides=2,
+                activation="relu",
+                padding="valid",
+            )
+        )
+        # we do not use a final sigmoid activation, since we use
+        # from_logits=True during training
+        self.segmask_output = keras.layers.TimeDistributed(
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:73]: segmask_output -> segmentation_mask_output

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+    """A Keras model implementing the Mask R-CNN architecture.
+
+    Mask R-CNN is a straightforward extension of Faster R-CNN, providing an
+    additional mask head that predicts segmentation masks. The constructor
+    therefore requires a FasterRCNN model as backend.
+
+    This model is compatible with Keras 3 only.
+
+    Args:
+         backbone: `keras.Model`. A FasterRCNN model that is used for
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:61]: backbone: `keras.Model`. A FasterRCNN model that is used for
            object detection. ==> backbone: `keras.Model`. A FasterRCNN backbone model.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+            object detection.
+        mask_head: (Optional) A `keras.Layer` that performs regression of
+            the segmentation masks.
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:64]: continue the arg description in the same line

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Thanks for the PR! Added some initial comments.

**Comment by divyashreepathihalli** [comment]
can you please add maskRCNN folder to this script? - https://github.com/keras-team/keras-cv/blob/ba2556c98d2e122c6b7c9b2f6a94097548a8ee8a/.kokoro/github/ubuntu/gpu/build.sh#L76

File: keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py [hunk 11]
```diff
@@ -0,0 +1,812 @@
...
+BOX_VARIANCE = [0.1, 0.1, 0.2, 0.2]
+
+
+# @keras_cv_export(
+#     [
+#         "keras_cv.models.FasterRCNN",
+#         "keras_cv.models.object_detection.FasterRCNN",
+#     ]
+# )
+class FasterRCNN(Task):
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py:54]: This should subclass backbone class. And the whole model implementation need not be added here. This could just be a simple wrapper for the FasterRCNN backbone

File: keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py [hunk 12]
```diff
@@ -0,0 +1,812 @@
...
+    def __init__(
+        self,
+        backbone,
+        num_classes,
+        bounding_box_format,
+        anchor_generator=None,
+        anchor_scales=[1],
+        anchor_aspect_ratios=[0.5, 1.0, 2.0],
+        feature_pyramid=None,
+        fpn_min_level=2,
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py:172]: Any arg that determines the model architecture - that is it would change the number of weights in the model, etc should not have a default value. These values will be initialized as saved in config.json when we `save_to_preset` to generate the presets.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+from keras_cv.src.api_export import keras_cv_export
+from keras_cv.src.backend import keras
+
+
+@keras_cv_export(
+    "keras_cv.models.mask_rcnn.MaskHead",
+    package="keras_cv.models.mask_rcnn",
+)
+class MaskHead(keras.layers.Layer):
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:23]: move the head layer to mask_rcnn.py file.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+
+    Args:
+        backbone: `keras.Model`. A FasterRCNN backbone model.
+        mask_head: (Optional) A `keras.Layer` that performs regression of
+            the segmentation masks. If not provided, a network with
+            2 convolutional layers, an upsampling layer and a class-specific
+            layer will be used.
+    """  # noqa: E501
+
+    def __init__(self, backbone, mask_head=None, **kwargs):
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:69]: mask_head should not be an arg. It should be part of teh default task model we have here.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
left some restructuring comments.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    """A Keras layer implementing the R-CNN Mask Head.
+
+    The architecture is adopted from Matterport's Mask R-CNN implementation
+    https://github.com/matterport/Mask_RCNN/blob/master/mrcnn/model.py.
+
+    Args:
+        num_classes: The number of object classes that are being detected.
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:30]: does this include background as one of the class. if that is the case it should be mentioned

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        conv_dims: (Optional) a list of integers specifying the number of
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:31]: if we are fllowing conventions from other models This should be renamed to `stackwise_num_conv_filters` and  `stackwise_num_deconv_filters`

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+                num_classes + 1,
+                kernel_size=1,
+                strides=1,
+                activation="linear",
+            )
+        )
+
+    def call(self, feature_map, training=False):
+        x = feature_map
+        for layer in self.layers:
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:85]: we cannot have for loops in the call method of the layer. The implementations within call has to be vectorized. This will slow down the execution.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 18]
```diff
@@ -0,0 +1,515 @@
...
+@keras_cv_export(
+    [
+        "keras_cv.models.MaskRCNN",
+        "keras_cv.models.object_detection.MaskRCNN",
+    ]
+)
+class MaskRCNN(Task):
+    """A Keras model implementing the Mask R-CNN architecture.
+
+    Mask R-CNN is a straightforward extension of Faster R-CNN, providing an
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:55]: nit: remove straightforward

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 19]
```diff
@@ -0,0 +1,515 @@
...
+            the segmentation masks. If not provided, a network with
+            2 convolutional layers, an upsampling layer and a class-specific
+            layer will be used.
+    """  # noqa: E501
+
+    def __init__(self, backbone, mask_head=None, **kwargs):
+        mask_head = mask_head or MaskHead(
+            backbone.num_classes, name="mask_head"
+        )
+
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:73]: remove unnecessary blank lines

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+        images = backbone.input
+        feature_map = backbone.get_layer("roi_align").output
+
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:76]: remove unnecessary blank lines

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 21]
```diff
@@ -0,0 +1,515 @@
...
+        rpn_classification_loss=None,
+        box_loss=None,
+        classification_loss=None,
+        mask_loss=None,
+        weight_decay=0.0001,
+        loss=None,
+        metrics=None,
+        **kwargs,
+    ):
+        if loss is not None:
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:113]: add comments to mention section of code computing loss

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py [hunk 22]
```diff
@@ -0,0 +1,515 @@
...
+            raise ValueError(
+                "`MaskRCNN` does not accept a `loss` to `compile()`. "
```
Comment (divyashreepathihalli) [inline keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:115]: I am confused here. Please explain.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

File: keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py (hunk 12 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/faster_rcnn_backbone.py:172]: This really should be changed in the faster RCNN.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py (hunk 13 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:23]: Done.This is inconsistent with Faster RCNN and other models, though.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 14 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:69]: Done.Also inconsistent with Faster RCNN.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py (hunk 15 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:30]: Done.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py (hunk 16 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:31]: I have changed to `stackwise_num_conv_filters` and `num_deconv_filters`. stackwise doesn't make sense for `num_deconv_filters` since there is no stack.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_head.py (hunk 17 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_head.py:85]: I think it is not possible to vectorize successive layer invocations.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 18 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:55]: Done.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 19 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:73]: Done.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 20 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:76]: Done.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 21 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:113]: Done.

File: keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py (hunk 22 above)
Comment (laxmareddyp) [reply keras_cv/src/models/object_detection/mask_rcnn/mask_rcnn.py:115]: Usually the `compile` method of keras model takes a `loss` argument. In this case, we have multiple losses(`box_loss`, `classification_loss`, and'mask_loss'). We cannot accept the provided `loss` argument and have different arguments instead. This has been adopted from Faster RCNN.
